Objects/listobject.c

```diff
@@ -2649,6 +2675,13 @@ list___init___impl(PyListObject *self, PyObject *iterable)
         (void)_list_clear(self);
     }
     if (iterable != NULL) {
+        Py_ssize_t iter_len = PyObject_Length(iterable);
```

Comment:
To reduce or avoid the overhead of the worst case, you can avoid this call if _PyObject_HasLen() is false.

This change doesn't require to implement a new _PyObject_ProbeLength().

Reply:
Updated run of first experiment using `_PyObject_HasLen()`:

```
Mean +- std dev: [old] 795 ns +- 25 ns -> [new] 776 ns +- 25 ns: 1.02x faster (-6%)
```